providers/darwin/arch_darwin.go

```diff
@@ -35,3 +41,37 @@ func Architecture() (string, error) {
 
 	return arch, nil
 }
+
+func NativeArchitecture() (string, error) {
+	processArch, err := Architecture()
+	if err != nil {
+		return "", err
+	}
+
+	// https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment
+
+	data, err := unix.Sysctl(procTranslated)
+	if err != nil {
+		// macos without Rosetta installed doesn't have sysctl.proc_translated
+		if os.IsNotExist(err) {
+			return processArch, nil
+		}
+		return "", fmt.Errorf("failed to read sysctl.proc_translated: %w", err)
+	}
+
+	nativeArch := ""
+	translated := data[0]
+
+	switch translated {
+	case 0:
+		nativeArch = processArch
+	case 1:
+		// Rosetta 2 is supported only on Apple silicon
+		if processArch == archIntel {
+			nativeArch = archApple
+		}
+	default:
+	}
+
+	return nativeArch, nil
```

Comment:
Using `SysctlUint32` would be more clear.

```suggestion
	translated, err := unix.SysctlUint32(procTranslated)
	if err != nil {
		// macOS without Rosetta installed doesn't have sysctl.proc_translated
		if os.IsNotExist(err) {
			return processArch, nil
		}
		return "", fmt.Errorf("failed to read %s: %w", procTranslated, err)
	}

	if translated > 0 {
		// Rosetta 2 is supported only on Apple silicon (arm64).
		return archApple, nil
	}

	return processArch, nil
```

Reply:
the syscall can return -1 which would mean we don't really know what's the native architecture, but thanks for the tip of `SysctlUint32`

Reply:
fixed 6a0c5210b6bb6f8a3612d7418fe75e48ce222a5f